### PR TITLE
reflection: move signature union-splitting logic under the control of inference

### PR DESCRIPTION
**Pull request by vtjnash**
It just makes more sense and is more useful for inference to have control over the union-splitting.

edit: I split a function, so https://github.com/JuliaLang/julia/pull/22144/files?w=1 is a bit easier to read

### CODE REVIEW

File: test/replcompletions.jl [hunk 1]
```diff
@@ -457,7 +457,6 @@ try
     end
 finally
     @test pop!(LOAD_PATH) == path
-    rm(path, recursive=true)
```
Comment (tkelman) [inline test/replcompletions.jl]: the parent dirs get created. and what does this have to do with the rest of this pr?

Comment (vtjnash) [inline test/replcompletions.jl]: This test is literally for the case "even though the path do no exist" (sic)

It doesn't, but it was tripping up my attempts at testing locally so it ended up on all my branches today.

Comment (tkelman) [inline test/replcompletions.jl]: only the first call, before mkpath gets run, and files get created under it. something else was happening in your case, open a separate issue instead of putting it into multiple unrelated pr's please

File: base/inference.jl [hunk 2]
```diff
@@ -1282,136 +1308,165 @@ function abstract_call_gf_by_type(f::ANY, atype::ANY, sv::InferenceState)
...
-        return Any
+    splitunions = 1 < countunionsplit(argtypes) <= sv.params.MAX_UNION_SPLITTING
+    if splitunions
+        splitsigs = switchtupleunion(argtype)
+        applicable = Any[]
+        for sig_n in splitsigs
+            xapplicable = _methods_by_ftype(sig_n, sv.params.MAX_METHODS, sv.params.world, min_valid, max_valid)
+            xapplicable === false && return Any
+            append!(applicable, xapplicable)
+        end
```
Comment (martinholters) [inline base/inference.jl]: Is it worth to keep track of how many different methods the entries in `applicable` actually refer to and bail out if that is more than `sv.params.MAX_METHODS`, too?

Comment (vtjnash) [inline base/inference.jl]: probably, for now I'm just moving the existing logic rather than trying to improve on it much

Comment (martinholters) [inline base/inference.jl]: Makes sense.

**Comment by martinholters** [comment]
:+1: Makes the code clearer by moving things where they belong. (Also a good idea to break  #21933 into smaller, more digestible pieces.)
